draft-ietf-quic-http.md

```diff
@@ -1102,18 +1109,33 @@ that can be prioritized are:
 
 Taken together, the dependencies across all prioritized elements in a connection
 form a dependency tree. An element can depend on another element or on the root
-of the tree. A reference to an element which is no longer in the tree is treated
-as a reference to the root of the tree. The structure of the dependency tree
-changes as PRIORITY frames modify the dependency links between prioritized
-elements.
-
-Due to reordering between streams, an element can also be prioritized which is
-not yet in the tree. Such elements are added to the tree with the requested
-priority.
+of the tree.  The tree also contains a dedicated placeholder with index "-1"
+which depends on the root of the tree with a weight of zero.  This placeholder
+cannot be reprioritized.  The structure of the dependency tree changes as
+PRIORITY frames modify the dependency links between other prioritized elements.
+
+All dependent streams are allocated an integer weight between 0 and 255
+(inclusive).
+
+Streams with the same parent SHOULD be allocated resources proportionally based
+on their weight.  Thus, if stream B depends on stream A with weight 4, stream C
+depends on stream A with weight 12, and no progress can be made on stream A,
+stream B ideally receives one-third of the resources allocated to stream C.
+
+A weight of zero indicates that no resources should be allocated to the
+indicated stream if progress can be made on other streams dependent on the same
+parent.
+
+A reference to an element which is no longer in the tree is treated as a
```

Comment:
If I read this correctly, if I reference stream 10 in a priority frame for stream 100 and you've removed all state for 10, 100 gets no resources whatsoever until all other requests finish?

That seems awful, so I must be missing something.

Reply:
From [Section 5.3.2](https://quicwg.org/base-drafts/draft-ietf-quic-http.html#priority-tree-maintenance):
> For prioritization purposes, a node in the tree is considered “inactive” when the corresponding stream has been closed for at least two round-trip times (using any reasonable estimate available on the server).

and

> Clients SHOULD assume the server is actively performing such pruning and SHOULD NOT declare a dependency on a stream it knows to have been closed.

So, yes, it is awful.  Is it more awful than saying it's the most important thing in the system because it parents directly to the root?  That's seriously context-dependent.  I'm inclined to minimize harm from a lost/delayed PRIORITY frame over minimizing harm from a misbehaving client.

Reply:
This is the main crux and why I proposed adding the orphan placeholder in the first place. Imo, it makes more sense for stream 100 to get no bandwidth, than to make it part of the root and thus make it one of the most important resources in the tree (even if it just has a weight of 16).